libcloud/compute/drivers/gridscale.py

```diff
@@ -0,0 +1,977 @@
+        for key, value in self._get_response_dict(result).items():
+            volume = self._to_volume(value)
+            volumes.append(volume)
+        return sorted(volumes, key=lambda sort: sort.extra['create_time'])
+
+    def ex_list_networks(self):
+        """
+        List all networks.
+
+        :return: List of objects.
+        :rtype: ``list`` of :class:`.GridscaleNetwork`
+        """
+        networks = []
+        result = self._sync_request(endpoint='objects/networks/')
+        for key, value in self._get_response_dict(result).items():
+            network = self._to_network(value)
+            networks.append(network)
+        return sorted(networks, key=lambda sort: sort.create_time)
+
+    def list_volume_snapshots(self, volume):
+        """
+        Lists all snapshots for storage volume.
+
+        :param volume: storage the snapshot is attached to
+        :type volume: :class:`.StorageVolume`
+
+        :return: Snapshots
+        :rtype: ``list`` of :class:`.VolumeSnapshot`
+        """
+        snapshots = []
+        result = self._sync_request(
+            endpoint='objects/storages/'
+                     '{}/snapshots'.format(volume.id))
+        for key, value in self._get_response_dict(result).items():
+            snapshot = self._to_volume_snapshot(value)
+            snapshots.append(snapshot)
+        return sorted(snapshots, key=lambda snapshot: snapshot.created)
+
+    def ex_list_ips(self):
+        """
+        Lists all IPs available.
+
+        :return: List of IP objects.
+        :rtype: ``list`` of :class:`.GridscaleIp`
+        """
+        ips = []
+        result = self._sync_request(endpoint='objects/ips/')
+        for key, value in self._get_response_dict(result).items():
+            ip = self._to_ip(value)
+            ips.append(ip)
+        return ips
+
+    def list_images(self):
+        """
+        List images.
+
+        :return: List of node image objects
+        :rtype: ``list`` of :class:`.NodeImage`
+        """
+        templates = []
+        result = self._sync_request(endpoint='objects/templates')
+        for key, value in self._get_response_dict(result).items():
+            template = self._to_node_image(value)
+            templates.append(template)
+        return sorted(templates, key=lambda sort: sort.name)
+
+    def create_node(self, name, size, image, location, auth):
+        """
+        Create a simple node  with a name, cores, memory at the designated
+        location.
+
+        :param name: Name of the server.
+        :type name: ``str``
+
+        :param size: Nodesize object.
+        :type size: :class:`.NodeSize`
+
+        :param image: OS image to attach to the storage.
+        :type image: :class:`.GridscaleTemplate`
+
+        :param location: The data center to create a node in.
+        :type location: :class:`.NodeLocation`
+
+        :param auth: sshkey uuid.
+        :type auth: :class:`.NodeAuthSSHKey`
```

Comment:
It looks like the ``auth`` argument is actually a string which contains an ID (UUID) of a SSH key which needs to be available on the account?

Reply:
If that's indeed the case, this method should take ``ex_ssh_key_ids`` argument instead.

Right now it looks like it actually takes ssh key id, but it incorrectly uses ``NodeAuthSSHKey`` class which expects a public key which is to be installed on the server.

Here is an example of using ``ex_ssh_key_ids`` argument - https://github.com/apache/libcloud/blob/trunk/libcloud/compute/drivers/digitalocean.py#L132.

Eventually, we could also add new ``NodeAuthSSHKeyId`` class (or similar) and make it part of a standard API and compatible with "auth" argument.

Reply:
The ssh key UUID is required to be able to attach an image to a node, while in the process of creating this node.

Reply:
Yes, but you should utilize ``ex_ssk_key_ids`` argument for that and not ``auth`` one.

Reply:
To clarify - ``ex_ssk_key_ids`` argument contains a list of key ids which are to be attached to the node upon creation.

On the other hand, ``NodeAuthSSHKey`` instance and the ``auth`` argument should contain a raw public key which is to be installed on the node.

Reply:
I've noticed it when i looked over the example you provided, this was very understandable.
thanks for the short summary :) .

Reply:
Also, I had a look at the API docs and I couldn't find anything about ``shkeyids`` attribute for the image templates.

I assume that's implementation details and that's how images are implemented?

Reply:
Thats right. There are a few things the API doesn't show just yet but will be in the future.